docs/sdk-spec.markdown

```diff
@@ -0,0 +1,256 @@
+Etcher SDK
+==========
+
+This document describes the design and interface of the Etcher SDK.
+
+Preface
+-------
+
+Etcher does a traditional type of image flashing, which is to simply copy the
+image byte-by-byte (with some exceptions, like omitting certain blocks when
+using block maps) to a destination device (like an SD Card).
+
+We discovered deviations of this flashing "mode," like to flash to a board's
+internal memory, and noticed a pattern: they all need two "actions":
+
+- Scanning for a set of destinations that are meaningful to the flashing mode
+- Flashing an image to one or more destinations found in the scanning phase
+
+In the future, there may be extra actions such as "configure" and "manage",
+which modes may decide to give or not.
+
+This specification describes a design for an Etcher SDK that supports different
+types of flashing modes.
+
+API
+---
+
+The Etcher SDK will expose this API:
+
+### `String[] etcher.SUPPORTED_MODES`
+
+An array of the modes this particular version of the Etcher SDK supports.
+
+### `EventEmitter etcher.image.fromURL(String url)`
+
+This function creates an *image* object out of an image URL. The implementation
+will at least support HTTP, HTTPS, and `file://`.
+
+- `String url`: The url to the image file
+
+The returned `EventEmitter` emits these events:
+
+- `done`: The operation was successful, and the handler argument is the image
+  object
+- `progress` (optional): The progress details, useful when constructing an
+  image object for a certain image will take more time
+- `error`: An error occurred
+
+### `EventEmitter etcher.image.fromFile(String path)`
+
+This function is like calling `etcher.image.fromURL()` with a `file://`
+protocol.
+
+### `String[] etcher.getSupportedModesForImage(Image image)`
+
+This function returns a list of supported modes for a particular image object.
+
+### `Promise<Destination[]> etcher.scan(Mode mode, Object options)`
+
+This function scans for destinations, passing the `options` object to the
+scanner's `.getAll()` function as-is.
+
+If the scanner singleton is not running, this function will yield back an empty
+set of destinations.
+
+### `EventEmitter etcher.flash(Mode mode, Image image, Destination[] destinations, Object options)`
+
+This function flashes an image to a set of destinations using a certain mode.
```

Comment:
`ble` stands for `bluetooth`  - they all use the bluetooth radio, but they all have very different update protocols. for micros `mode` would basically mean `device type`